django_echosign/tests/test_adobe_client.py

```diff
@@ -36,3 +40,159 @@ def test_oauth_create(mocker, adobe_oauth_session):
+    with pytest.raises(EchoSignException):
+        adobe_sign_client.upload_document(test_document)
+
+
+def test_should_create_signature(mocker, adobe_sign_client: EchoSignClient,
+                                 expected_participant, expected_headers):
+    mocked_post = mocker.patch('requests.post')
+    participants = [expected_participant]
+    adobe_sign_client \
+        .post_agreement(transient_document_id='test_doc_id',
+                        name='test_name',
+                        participants=participants,
+                        post_sign_redirect_url='http://post/sign/red/url',
+                        post_sign_redirect_delay=12,
+                        state='AUTHORING',
+                        extra_param_str='plop',
+                        extra_param_dict={'poney': 42},
+                        extra_param_list=[1, 2, 3])
+    expected_json = {
+        'fileInfos': [{
+            'transientDocumentId': 'test_doc_id'
+        }],
+        'name': 'test_name',
+        'participantSetsInfo': [expected_participant],
+        'signatureType': 'ESIGN',
+        'state': 'AUTHORING',
+        'postSignOption': {
+            "redirectDelay": 12,
+            "redirectUrl": 'http://post/sign/red/url'
+        },
+        'extra_param_str': 'plop',
+        'extra_param_dict': {'poney': 42},
+        'extra_param_list': [1, 2, 3]
+    }
+    # By default the assert call checks the last call
+    mocked_post.assert_called_with('http://test/api/rest/v6/agreements',
+                                   headers=expected_headers,
+                                   json=expected_json)
+
+
+@pytest.mark.parametrize('error_code', (404, 500))
+def test_post_agreement_client_or_server_error(error_code, mocker,
+                                               adobe_sign_client):
+    response_with_client_error = Response()
+    response_with_client_error.status_code = error_code
+    expected_json_reply_error = {'code': str(error_code),
+                                 'message': 'error raison'}
+    mocker.patch.object(response_with_client_error, 'json',
+                        return_value=expected_json_reply_error)
+    mocker.patch('requests.post', return_value=response_with_client_error)
+    with pytest.raises(EchoSignException) as e:
+        adobe_sign_client.post_agreement('doc id', 'name', [], '-', '-')
+    assert expected_json_reply_error['code'] in str(e)
+    assert expected_json_reply_error['message'] in str(e)
+
+
+def test_get_agreements(mocker, adobe_sign_client, expected_headers):
+    mocked_get = mocker.patch('requests.get')
+    adobe_sign_client.get_agreements(page_size=12,
+                                     cursor=44,
+                                     extra_param_test={'test': 1})
+    expected_params = {'pageSize': 12,
+                       'cursor': 44,
+                       'extra_param_test': {'test': 1}}
+    mocked_get.assert_called_with('http://test/api/rest/v6/agreements',
```

Comment:
I don't like `assert_called_with` because if there's a typo (`asert_called_with`), it will pass nonetheless. I'd rather have an explicit assertion like:

```python
assert mocked_get.mock_calls == [...]
```

Reply:
```python
>>> from unittest import mock
>>> m = mock.MagicMock()
>>> m.yay(1, a=2)
<MagicMock name='mock.yay()' id='140578457536944'>
>>> m.call_args_list
[]
>>> m.mock_calls
[call.yay(1, a=2)]

>>> m.yay.call_args_list
[call(1, a=2)]
>>> m.yay.mock_calls
[call(1, a=2)]
```

So with mock_calls you see ALL calls to mock objects and sub-objects, with call_args_list you see the calls to the object only.
`call` can be accessed as `mocker.call`.

Reply:
I agree mock are awful :) and I will replace assert_called_with wit explicit assertions. 
But, on the other side, assert with call_args_list or call_args implies a loss of readability.
Furthermore, even if we don't apply TDD principles, to avoid to write a test that inadvertently passes we must write failing unit test at first run.